docs/lang/articles/advanced/layout.md

```diff
@@ -318,53 +301,34 @@ x = ti.field(dtype=ti.f32)
 fb1.dense(ti.ij, (5, 5)).place(x)
 fb1_snode_tree = fb1.finalize()  # Finalizes the FieldsBuilder and returns a SNodeTree
 func(x)
+...
+fb1_snode_tree.destroy() # Destruction
 
 fb2 = ti.FieldsBuilder()
 y = ti.field(dtype=ti.f32)
 fb2.dense(ti.i, 5).place(y)
 fb2_snode_tree = fb2.finalize()  # Finalizes the FieldsBuilder and returns a SNodeTree
 func(y)
+...
+fb2_snode_tree.destroy() # Destruction
 ```
+Actually, the above demonstrated `ti.root` statements are implemented with `FieldsBuilder`, despite that `ti.root` has the capability to automatically manage memory allocations and recycling.
 
-In fact, `ti.root` is implemented by `FieldsBuilder` implicitly, so you can
-allocate the fields directly under `ti.root`:
-```python
-import taichi as ti
-ti.init()  # Implicitly: ti.root = ti.FieldsBuilder()
+### Packed mode
 
-@ti.kernel
-def func(v: ti.template()):
-    for I in ti.grouped(v):
-        v[I] += 1
+By default, Taichi implicitly fits a field in a larger buffer with power-of-two dimensions. We take the power-of-two padding convention as it is widely adopted in computer graphics. The design enables fast indexing with bitwise arithmetics and better memory address alignment, while trades off memory occupations especially for large irregular fields.
 
-x = ti.field(dtype=ti.f32)
-ti.root.dense(ti.ij, (5, 5)).place(x)
-func(x)  # Automatically calls ti.root.finalize()
-# Implicitly: ti.root = ti.FieldsBuilder()
+For example, a `(18, 65)` field is materialized with a `(32, 128)` buffer, which is acceptable. As field size grows, the padding stragtegy can be exaggeratedly unbearable: `(129, 6553600)` will be padded to `(256, 6335600)`, which is a waste of the precious memory capacity. Therefore, Taichi provides the optional packed mode to allocate buffer that tightly fits the requested field shape. It is especially useful when memory usage is a major concern.
 
-y = ti.field(dtype=ti.f32)
-ti.root.dense(ti.i, 5).place(y)
-func(y)  # Automatically calls ti.root.finalize()
+To leverage the packed mode, spcifify `packed` in `ti.init()` argument:
+```python
+ti.init()  # default: packed=False
+a = ti.field(ti.i32, shape=(18, 65))  # padded to (32, 128)
 ```
 
-Furthermore, if you don't want to use the fields under a certain `SNodeTree`
-anymore, you could call the `destroy()` method on the finalized `SNodeTree`
-object, which will recycle its memory into the memory pool:
-
-```py
-import taichi as ti
-ti.init()
-
-@ti.kernel
-def func(v: ti.template()):
-    for I in ti.grouped(v):
-        v[I] += 1
-
-fb = ti.FieldsBuilder()
-x = ti.field(dtype=ti.f32)
-fb.dense(ti.ij, (5, 5)).place(x)
-fb_snode_tree = fb.finalize()  # Finalizes the FieldsBuilder and returns a SNodeTree
-func(x)
-
-fb_snode_tree.destroy()  # x cannot be used anymore
+```python
+ti.init(packed=True)
+a = ti.field(ti.i32, shape=(18, 65))  # no padding
 ```
+
+You might obeserve slight performance regression, which is not a big concern in most cases.
```

Comment:
this is not a good sentence to end this page ;// suggest to detail the situation and provide some suggestions if possible?

Reply:
Yes it's a hard brake here. We have some options: 1) a conclusion 2) some more examples and coding suggestions. 3) I think the contents other than memory layout are way too skinny. Is there some other uncovered topics regarding field?

Reply:
I vote for a section conclusion here. This sentence is still for the Packed mode section?  The performance regression is due to the lack of padding of  the memory. I would rephrase it to something to echo the first paragraph: "Since packed mode trades memory utilization for fast bitwise arithmetic operations and address alignment, you might observe slight performance regression when enabled.". Just some suggestion ;/